### PR TITLE
Clone tensors to avoid torch.save bloat

### LINKED ISSUE #3303: [BUG] Size of saved model checkpoint becomes much larger after deepspeed.initialize when using deepspeed stage2
**Describe the bug**
Originally reported [here](https://github.com/huggingface/transformers/issues/22822). @stas00 @tjruwase

For some models, the size of model checkpoints saved by `model.save_prtrained()` becomes much larger after calling `deepspeed.initialize`. See examples below.

**To Reproduce**
```python
from transformers import AutoModelForCausalLM
import deepspeed

ds_config = {
    "optimizer": {
        "type": "AdamW",
    },
    "zero_optimization": {
        "stage": 2,
        "offload_optimizer": {
            "device": "cpu",
            "pin_memory": True
        },
        "allgather_partitions": True,
        "allgather_bucket_size": 2e8,
        "overlap_comm": True,
        "reduce_scatter": True,
        "reduce_bucket_size": 2e8,
        "contiguous_gradients": True
    },
    "offload_optimizer": {
        "device": "cpu",
        "pin_memory": True
    },
    "train_batch_size": 1,
    "train_micro_batch_size_per_gpu": 1
}

model = AutoModelForCausalLM.from_pretrained("decapoda-research/llama-7b-hf")
model.save_pretrained("before")
deepspeed_engine, _, _, _ = deepspeed.initialize(model=model, config_params=ds_config)
deepspeed_engine.module.save_pretrained("after")
```

File sizes:

```bash
du -a -h --max-depth=1 before/
512     before/config.json
32K     before/pytorch_model.bin.index.json
9.2G    before/pytorch_model-00001-of-00003.bin
9.3G    before/pytorch_model-00002-of-00003.bin
6.7G    before/pytorch_model-00003-of-00003.bin
512     before/generation_config.json
26G     before/

du -a -h --max-depth=1 after/
512     after/config.json
32K     after/pytorch_model.bin.index.json
26G     after/pytorch_model-00001-of-00003.bin
26G     after/pytorch_model-00002-of-00003.bin
26G     after/pytorch_model-00003-of-00003.bin
512     after/generation_config.json
76G     after/
```

This issue is not always occurred, for example, `gpt2` does not have this problem. But I tested  `decapoda-research/llama-7b-hf`, and `decapoda-research/llama-13b-hf` have this issue.

This can be fixed by re-clone states before the saving:
```python
state_dict = deepspeed_engine.module.state_dict()
state_dict = type(state_dict)(
    {k: v.clone()
     for k,
     v in state_dict.items()})
deepspeed_engine.module.save_pretrained("after_fixed", state_dict=state_dict)
```

**Expected behavior**
The saved model size should be unchanged after `deepspeed.initialize`

**System info (please complete the following information):**
- deepspeed: 0.8.3
- transformers version: 4.28.0.dev0
- Platform: Linux-4.18.0-372.32.1.el8_6.x86_64-x86_64-with-glibc2.17
- Python version: 3.8.16
- Huggingface_hub version: 0.13.3
- Safetensors version: not installed
- PyTorch version (GPU?): 1.12.1+cu116 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: yes
- Using distributed or parallel set-up in script?: yes

### PR DESCRIPTION
**Pull request by tjruwase**
Fixes #3303

### CODE REVIEW

**Comment by tjruwase** [comment]
TODOs: 
 1. Docs
 2. Unit tests (?)

**Comment by tjruwase** [comment]
@stas00, please see docs https://deepspeed.readthedocs.io/en/rtd-staging/model-checkpointing.html#avoiding-zero-checkpoint-bloat

File: docs/code-docs/source/model-checkpointing.rst [hunk 1]
```diff
@@ -22,3 +22,24 @@ DeepSpeed provides routines for extracting fp32 weights from the saved ZeRO chec
...
+
+.. code-block:: python
+    ...
+    ds_config = {
+     ...
+    }
+    model = AutoModelForCausalLM.from_pretrained("facebook/opt-13b", torch_dtype=torch.float16)
+    ds_engine, _, _, _ = deepspeed.initialize(model=model, config_params=ds_config)
+    lean_state_dict = deepspeed.checkpoint.utils.clone_tensors_for_torch_save(ds_engine.module.state_dict())
+    ds_engine.module.save_pretrained("lean_after", state_dict=lean_state_dict)
```
Comment (stas00) [inline docs/code-docs/source/model-checkpointing.rst:45]:
```suggestion
Avoiding ZeRO Checkpoint Bloat
------------------------------
ZeRO stage 1 and 2 checkpoints created using ``torch.save()`` can sometimes be larger than expected. This bloat
is caused by the interaction of ZeRO's tensor flattening and torch's tensor `storage management https://pytorch.org/docs/stable/notes/serialization.html#preserve-storage-sharing`.
You can avoid this problem by using the ``clone_tensors_for_torch_save`` utility of DeepSpeed as illustrated below.

.. autofunction:: deepspeed.checkpoint.utils.clone_tensors_for_torch_save

The following code snippet illustrates this functionality:

.. code-block:: python

    ds_config = {
         ...
    }
    model = AutoModelForCausalLM.from_pretrained("facebook/opt-13b", torch_dtype=torch.float16)
    ds_engine, _, _, _ = deepspeed.initialize(model=model, config_params=ds_config)
    lean_state_dict = deepspeed.checkpoint.utils.clone_tensors_for_torch_save(ds_engine.module.state_dict())
    ds_engine.module.save_pretrained("lean_after", state_dict=lean_state_dict)
    ...

Comment (stas00) [inline docs/code-docs/source/model-checkpointing.rst:45]:
```suggestion
The following code snippet illustrates this functionality.

.. code-block:: python

    ds_config = {
     ...
    }
    model = AutoModelForCausalLM.from_pretrained("facebook/opt-13b", torch_dtype=torch.float16)
    ds_engine, _, _, _ = deepspeed.initialize(model=model, config_params=ds_config)
    lean_state_dict = deepspeed.checkpoint.utils.clone_tensors_for_torch_save(ds_engine.module.state_dict())
    ds_engine.module.save_pretrained("lean_after", state_dict=lean_state_dict)
    ...

Comment (stas00) [reply docs/code-docs/source/model-checkpointing.rst:45]: this should probably fix the formatting?

File: deepspeed/checkpoint/utils.py [hunk 2]
```diff
@@ -29,3 +30,31 @@ def get_layer_ckpt_name_for_rank(base_folder, layer_id, tp_rank):
...
+# See https://pytorch.org/docs/stable/notes/serialization.html#preserve-storage-sharing
+def clone_tensors_for_torch_save(item, device=torch.device('cpu')):
+    """
+    Returns a copy of input with all enclosed tensors replaced by clones on specified device.
+    Works on individual tensors, and tensors contained/nested in lists, tuples, and dicts.
+    Parameters:
+        item: tensor to clone or (possibly nested) container of tensors to clone.
+        device: target device
+    Returns:
+        copy of input with cloned tensors on target device
```
Comment (stas00) [inline deepspeed/checkpoint/utils.py:49]:
```suggestion
def clone_tensors_for_torch_save(item, device=torch.device('cpu')):
    """
    Returns a copy of ``item`` with all enclosed tensors replaced by clones on specified device.
    Works on individual tensors, and tensors contained/nested in lists, tuples, and dicts.
    
    Parameters:
        - ``item``: tensor to clone or (possibly nested) container of tensors to clone.
        - ``device``: target device (defaults to 'cpu')
        
    Returns:
        - copy of ``item`` with cloned tensors on target device
```

File: deepspeed/checkpoint/utils.py [hunk 3]
```diff
@@ -29,3 +30,31 @@ def get_layer_ckpt_name_for_rank(base_folder, layer_id, tp_rank):
...
+    """
+    if torch.is_tensor(item):
+        return item.clone().to(device)
```
Comment (stas00) [inline deepspeed/checkpoint/utils.py:52]: probably need to call `.detach()` as well just in case.

Comment (stas00) [reply deepspeed/checkpoint/utils.py:52]:
```suggestion
        return item.clone().detach().to(device)
```

Comment (stas00) [reply deepspeed/checkpoint/utils.py:52]: may be let's not do anything and let the user do any `detach` if they need it to?

**Comment by stas00** [comment]
Looking good now!

**Review by ShijieZZZZ — APPROVED** [review]
